File: nodejs/linked-list-components-817.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
'use strict';
/**
 * @param {ListNode} head
 * @param {number[]} G
 * @return {number}
 */
var numComponents = function (head, G) {
    let set = new Set();
    for (let v of G) {
        set.add(v);
    }

    let connected = false;
    let count = 0;
    while (head !== null) {
        if (set.has(head.val)) {
            if (!connected) {
                count++;
                connected = true;
            }
        }
        else {
            connected = false;
        }
        head = head.next;
    }
    return count;
};

function ListNode(val) {
    this.val = val;
    this.next = null;
}


let node = new ListNode(44);
node.next = new ListNode(33);
node.next.next = new ListNode(22);
node.next.next.next = new ListNode(11);

numComponents(node,[]);

File: golang/linkedlistcomponents817/linked-list-components-817.go
package main

import "fmt"

func main() {

	head := &ListNode{Val: 44, Next: &ListNode{Val: 23, Next: &ListNode{Val: 22, Next: &ListNode{Val: 11, Next: nil}}}}

	G := [...]int{11, 22, 44, 23}

	numComponents(head, G[:])

}
func numComponents(head *ListNode, G []int) int {
	m := make(map[int]*ListNode)

	iNode := head
	for iNode != nil {
		m[iNode.Val] = iNode
		iNode = iNode.Next
	}

	count := 0

	mg := make(map[int]*ListNode, len(G))
	for _, v := range G {
		mg[v] = m[v]
	}

	for _, v := range G {
		next := mg[v].Next
		if next == nil {
			count++
		} else if _, ok := mg[next.Val]; !ok {
			count++
		}
	}

	fmt.Printf("Count is:\t%d\n", count)
	return count
}

type ListNode struct {
	Val  int
	Next *ListNode
}
